### PR TITLE
Assign a UUID as the logical ID (and return it) of a newly created FHIR resource

### LINKED ISSUE #784: Design and implementation for generation of UUID by FHIR Engine
**Is your feature request related to a problem? Please describe.**
How do we handle resource IDs by the FHIR SDK. This has implications for how we can do the Sync from the Android client to the FHIR Server and also resolve potential ID conflict issues

See [this deck](https://docs.google.com/presentation/d/1xx1mt8wjktVanrjQk3YQe0jmfyigTcQLM24kxv6v79A/edit#slide=id.gf01cad751a_0_5) for exploration of the different approaches

**Describe the solution you'd like**
Use PUT to create new resources on the server side. The SDK does not enforce (but strongly recommends) UUID creation (option 3)

* The SDK should be able to specify UUIDs for resources created by the client application 
* Resources created on the SDK side should be able to be sync'd to the FHIR server using PUT (not POST)
* The FHIR SDK will not enforce, but strongly recommend UUID creation - **default behaviour will be to create the UUIDs providing implementers the ability to override**

**Describe alternatives you've considered**
See this deck for detailed alternatives. In summary these include:
1. Server sets the resource IDs on POSTing the new resources
2. FHIR SDK enforces the creation of UUIDs

**Additional context**
1. UUIDs will only be generated for newly created resources.
2. This conforms to the standard according to the [FHIR spec](https://www.hl7.org/fhir/http.html#upsert). The FHIR SDK is effectively an extension of the FHIR server 
3. There will be server side requirements that will need to be enabled to support this. This will be documented as part of an ImplementationGuide

### PR DESCRIPTION
**Pull request by jingtang10**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #784 

**Description**
Assign a UUID as the logical ID (and return it) of a newly created FHIR resource.

Additionally
- Remove the incorrect `@Transatction` annotation
- Rename `save` to `create` to distinguish it from `update`
- Remove the generic type parameter `R` for `create` and `update`
- Make `update` work with more than one resource

**Alternative(s) considered**
NA

**Type**
Choose one: Feature

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

**Comment by Tarun-Bhardwaj** [comment]
@aditya-07 , could you review this PR please?

**Review by aditya-07 — COMMENTED** [review]
We should add some basic tests to make sure the api works as expected when creating resources with or without id.

File: engine/src/androidTest/java/com/google/android/fhir/db/impl/DatabaseImplTest.kt [hunk 1]
```diff
@@ -90,7 +90,7 @@ class DatabaseImplTest {
   private val testingUtils = TestingUtils(services.parser)
   private val database = services.database
 
-  @Before fun setUp() = runBlocking { database.insert(TEST_PATIENT_1) }
+  @Before fun setUp(): Unit = runBlocking { database.insert(TEST_PATIENT_1) }
```
Comment (jingtang10) [inline engine/src/androidTest/java/com/google/android/fhir/db/impl/DatabaseImplTest.kt:93]: why is this needed?

```suggestion
  @Before fun setUp() = runBlocking { database.insert(TEST_PATIENT_1) }
```

Comment (aditya-07) [reply engine/src/androidTest/java/com/google/android/fhir/db/impl/DatabaseImplTest.kt:93]: `@Before` functions should return void & `runBlocking` return the last line. The tests start failing if we don't explicitly mark it `Unit`.

File: engine/src/test/java/com/google/android/fhir/impl/FhirEngineImplTest.kt [hunk 2]
```diff
@@ -75,11 +97,58 @@ class FhirEngineImplTest {
...
+      }
+    fhirEngine.create(patient1)
+    val updatedPatient1 = patient1.copy().setGender(Enumerations.AdministrativeGender.FEMALE)
+    val updatedPatient2 = patient2.copy().setGender(Enumerations.AdministrativeGender.MALE)
+
+    assertThrows(ResourceNotFoundException::class.java) {
+      runBlocking { fhirEngine.update(updatedPatient1, updatedPatient2) }
+    }
+
+    testingUtils.assertResourceNotEquals(
```
Comment (stevenckngaa) [inline engine/src/test/java/com/google/android/fhir/impl/FhirEngineImplTest.kt:147]: Also assert updatedPatient2 not equals to the patient 2 stored in the database.

**Review by stevenckngaa — APPROVED** [review]

File: engine/src/test/java/com/google/android/fhir/impl/FhirEngineImplTest.kt (hunk 2 above)
Comment (aditya-07) [reply engine/src/test/java/com/google/android/fhir/impl/FhirEngineImplTest.kt:147]: We don't store patient 2 in the database. That's why `fhirEngine.update ` api throws `ResourceNotFoundException`.

File: engine/src/main/java/com/google/android/fhir/db/impl/DatabaseImpl.kt [hunk 3]
```diff
@@ -100,23 +100,27 @@ internal class DatabaseImpl(
...
-    db.withTransaction {
-      val oldResourceEntity = selectEntity(resource.javaClass, resource.logicalId)
-      resourceDao.update(resource)
-      localChangeDao.addUpdate(oldResourceEntity, resource)
-    }
+  override suspend fun <R : Resource> update(vararg resources: R) {
+    db.withTransaction { resources.forEach { update(it) } }
+  }
+
+  private suspend fun <R : Resource> update(resource: R) {
```
[the remark this answers is not in the document]
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/db/impl/DatabaseImpl.kt]: Done
